### DISCUSSION TITLE
Distorted mesh elements

### DISCUSSION THREAD

**Discussion by mauroarcidiacono** [neperfepx/neper#449] [Q&A] [answered]
Hello,

I am trying to model a representative volume element of a microstructure where some of the grains are laminar. I used the following commands:

`neper -T -n 30::from_morpho -id 4::1 -morpho "diameq:lognormal(1,0.1),sphericity:lognormal(0.145,0.03,1-x)::lamellar(w=file(lamwidth.txt),v=file(lamnormal.txt))" -reg 1 -sel 0.025 -oridescriptor e -format tess,ori -o Polycrystal_laminated`

`neper -M Polycrystal_laminated.tess -statelset vol -order 1 -cl 0.045 -meshqualexpr Odis -meshqualmin 1 -format inp,msh -o abq_input`

I was able to import the model to Abaqus but the problem that I have is that meshing some sets produce elements that are distorted. I attached a picture. Is there a way in which I can avoid having lamella that will generate distorted mesh elements? I thought about discarding the lamella that are under a certain volume, but I haven’t found a way to do this. I need to improve the quality of the mesh so I don’t have convergence issues.

![Distorted-lam-el](https://user-images.githubusercontent.com/76496290/164503821-5533278c-7028-4fc8-a63e-e8a69bb071b2.jpg)

Thank you very much in advance for your help.

**Comment by mkasemer** [comment, marked as answer]
I believe one issue may be the imposition of a self-defined small edge length (`-sel`). Have you attempted to try to use `rsel`, instead, or increasing the value for `sel`?

For the image you show above, I don't believe it is necessarily the issue that you highlighted with the white circle. Regularization likely did its job at that point and deleted a small edge to pin the blue lamellar region to the corner of the domain. I instead think it is likely due to other small edges---most visibly where the blue lamella touches the orange/brown grain at the right hand side. Increasing the length of small edges that you want to have deleted during regularization may allow for these artifacts to be removed, facilitating a more homogeneous mesh.

**Reply by rquey** [reply]
I will add that, as of version 4.4.1, there is `pos=optimal` (instead of the default `random`) in `-morpho lamellar`, which positions the lamellae (within their parent cells, along their normal) so that small/sliver lamellae are avoided. This will help a lot, I think. See [-morpho](https://neper.info/doc/neper_t.html#cmdoption-morpho).

**Reply by mauroarcidiacono** [reply]
Hi, thank you for your help and for answering so promptly. I have a few questions:

- Could you please explain how can I update Neper version to the latest release? I am currently running Neper version 4.2.0 (Gmsh version 4.7.1) and I believe that pos=optimal is not available. The result of running `neper -T -n 30::from_morpho -id 4::1 -morpho "diameq:lognormal(1,0.1),sphericity:lognormal(0.145,0.03,1-x)::lamellar(w=file(lamwidth.txt),v=file(lamnormal.txt),pos=optimal)" -reg 1 -sel 0.06 -oridescriptor e -format tess,ori -o Polycrystal_laminated` is: `Error  :     > Failed to process expression pos=value`. 

- If I use v=random, is it possible to print the normal vectors to the lamella planes (vx, vy, vz)? 

Thank you very much for your time and support.

**Reply by mauroarcidiacono** [reply]
Hi,

I want to leave a reply for those who have similar questions. The parameter pos=optimal was not available in version 4.2.0. I updated Neper to version 4.4.2-16 and now it is working. To update Neper, I did the following:

1. Download and extract the code from the Neper repository (I had a path length problem which I avoided by using 7-zip for the extraction).
2. Change directory to src inside of the Neper folder.
3. Create a build directory and cd to that directory. 
4. Execute the following commands in order: $ cmake .., $ sudo make, $ sudo make install.

I am currently using gmsh version 4.7.1. Regarding printing the normal vectors to the lamella planes, I did not find a way to do it. I am using a Python script to obtain that information from Abaqus.
